theano/scan_module/tests/test_scan.py

```diff
@@ -3967,8 +3971,16 @@ def onestep(x, x_tm4):
         updates = OrderedDict([(sharedvar, results[0][-1:])])
 
         f = theano.function([seq], results[1], updates=updates)
+
+        # This fails if scan uses wrongly the borrow flag
         assert numpy.all(exp_out == f(inp))
 
+        # This fails if Scan's infer_shape() is unable to remove the Scan
+        # node from the graph.
+        f_infershape = theano.function([seq], results[1].shape)
+        scan_nodes_infershape = scan_nodes_from_fct(f_infershape)
+        assert(len(scan_nodes_infershape) == 0)
```

Comment:
Is the ShapeFeature registered in FAST_COMPILE? If not, adding it could slow down the whole compilation phase. We should profile that.
I don't think we expect shape inference to be performed in FAST_COMPILE in general, right? So we could just force the mode here for the tests.